src/ray/common/status_or.h

```diff
@@ -0,0 +1,257 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <stdexcept>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+
+#include "absl/base/attributes.h"
+#include "ray/common/status.h"
+
+namespace ray {
+
+// TODO(hjiang): Introduce statusor related macros.
+template <typename T>
+class StatusOr {
+ public:
+  StatusOr() noexcept = default;
+  // NOLINTNEXTLINE(runtime/explicit)
+  StatusOr(Status status) : status_(std::move(status)) {}
+  // NOLINTNEXTLINE(runtime/explicit)
+  StatusOr(const T &data) : status_(Status::OK()), data_(data) {}
+  // NOLINTNEXTLINE(runtime/explicit)
+  StatusOr(T &&data) : status_(Status::OK()), data_(std::move(data)) {}
+
+  template <typename... Args>
+  explicit StatusOr(std::in_place_t ip, Args &&...args)
+      : data_(std::forward<Args>(args)...) {}
+
+  template <typename U>
+  StatusOr(const StatusOr<U> &status_or) {
+    if (status_or.ok()) {
+      data_ = status_or.value();
+      status_ = Status::OK();
+    } else {
+      status_ = status_or.status();
+    }
+  }
+
+  template <typename U>
+  StatusOr(StatusOr<U> &&status_or) {
+    if (status_or.ok()) {
+      data_ = std::move(status_or).value();
+      status_ = Status::OK();
+    } else {
+      status_ = std::move(status_or).status();
+    }
+  }
+
+  StatusOr(const StatusOr &) = default;
+  StatusOr &operator=(const StatusOr &) = default;
+
+  StatusOr(StatusOr &&) noexcept(std::is_nothrow_move_constructible_v<T>) = default;
+  StatusOr &operator=(StatusOr &&) noexcept(std::is_nothrow_move_assignable_v<T>) =
+      default;
+
+  ~StatusOr() noexcept(std::is_nothrow_destructible_v<T>) = default;
+
+  // Returns whether or not this `ray::StatusOr<T>` holds a `T` value. This
+  // member function is analogous to `ray::Status::ok()` and should be used
+  // similarly to check the status of return values.
+  //
+  // Example:
+  //
+  // StatusOr<Foo> result = DoBigCalculationThatCouldFail();
+  // if (result.ok()) {
+  //    // Handle result
+  // else {
+  //    // Handle error
+  // }
+  bool ok() const { return status_.ok(); }
+  bool has_value() const { return ok(); }
```

Comment:
i know has_value is common, but don't see any point of having both, just makes things less consistent across the codebase

Reply:
ok, removed

Reply:
Second thought, `std::expect` actually supports both boolean casting and `has_value`
https://en.cppreference.com/w/cpp/utility/expected/operator_bool